internal/pkg/postprocessor/extractor/html_document.go

```diff
@@ -0,0 +1,148 @@
+package extractor
+
+import (
+	"bytes"
+	"io"
+	"net/url"
+	"unicode/utf8"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/internetarchive/Zeno/internal/pkg/log"
+	"github.com/internetarchive/Zeno/pkg/models"
+	"golang.org/x/net/html/charset"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/htmlindex"
+	"golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/transform"
+)
+
+var htmldocLogger = log.NewFieldedLogger(&log.Fields{
+	"component": "postprocessor.extractor.html_document",
+})
+
+// This function is a modified version of the charset.NewReader() function
+// Returns additional [enc, name, certain] so we can get the [enc] encoding info to encode url query components correctly
+func charsetNewReader(r io.Reader, contentType string) (io.Reader, error, encoding.Encoding, string, bool) {
+	preview := make([]byte, 1024)
+	n, err := io.ReadFull(r, preview)
+	switch {
+	case err == io.ErrUnexpectedEOF:
+		preview = preview[:n]
+		r = bytes.NewReader(preview)
+	case err != nil:
+		return nil, err, nil, "", false
+	default:
+		r = io.MultiReader(bytes.NewReader(preview), r)
+	}
+
+	var enc encoding.Encoding
+	htmlEnc, name, certain := charset.DetermineEncoding(preview, contentType)
+	if htmlEnc != nil && htmlEnc != encoding.Nop {
+		// The htmlEnc will use HTML escape sequences for runes that are not supported by the character set.
+		// To get the original unwrapped encoding, we need to use htmlindex.
+		enc, err = htmlindex.Get(name)
+		if err != nil { // This should never happen
+			htmldocLogger.Error("unable to get encoding from htmlindex", "err", err.Error(), "name", name)
+			return r, err, nil, "", false
+		}
+
+		r = transform.NewReader(r, enc.NewDecoder())
+	}
+
+	return r, nil, enc, name, certain
+}
+
+// TransformDocument transforms the document of a URL by detecting its encoding and creating a utf-8 goquery document.
+func TransformDocument(u *models.URL) (doc *goquery.Document, err error) {
+	u.DocumentTransformMux.Lock()
```

Comment:
Is anything else expected to be using the document while it's being transformed? I can't see another use of the mutex?

Reply:
I had previously assumed that the outlink extrator and asset extractor would run simultaneously. just checked, they are executed sequentially and indeed no locks are needed here.